### PR TITLE
Fix codoon and Upload tcx_to_garmin

### LINKED ISSUE #645: 使用--with-tcx参数获取咕咚数据，导出来的TCX文件导入Garmin还是会提示报错无法导入
如题

### PR DESCRIPTION
**Pull request by Vensent**
1. Fix the tcx format in codoon_sync.py
2. Add tcx_sync_garmin.py
3. Add oppo_sync.py, while currently not work.

Closes #645

### CODE REVIEW

**Comment by yihong0618** [comment]
nice pr will take a look tonight

File: .github/workflows/run_data_sync.yml [hunk 1]
```diff
@@ -187,6 +195,11 @@ jobs:
         run: |
           python run_page/tulipsport_sync.py ${{ secrets.TULIPSPORT_TOKEN }} --with-gpx
 
+      - name: Run sync Oppo heytap script, note currently this script is not worked
+        if: env.RUN_TYPE == 'oppo'
+        run: |
+            python run_page/oppo_sync.py ${{ secrets.OPPO_ID }} ${{ secrets.OPPO_CLIENT_SECRET }} ${{ secrets.OPPO_CLIENT_REFRESH_TOKEN }} --with-tcx
```
Comment (yihong0618) [inline .github/workflows/run_data_sync.yml:201]: we do not support gpx right?

File: run_page/codoon_sync.py [hunk 2]
```diff
@@ -217,9 +243,10 @@ def tcx_output(fit_array, run_data):
             tp.append(altitude_meters)
 
     # write to TCX file
-    tree.write(
-        TCX_FOLDER + "/" + fit_id + ".tcx", encoding="utf-8", xml_declaration=True
-    )
+    xmlstr = minidom.parseString(ET.tostring(training_center_database)).toprettyxml(indent="  ", encoding="UTF-8")
+    with open(TCX_FOLDER + "/" + fit_id + ".tcx", "w") as f:
+        f.write(str(xmlstr.decode("UTF-8")))
+        f.close()
```
Comment (yihong0618) [inline run_page/codoon_sync.py:249]: do not need close here

**Review by yihong0618 — COMMENTED** [review]
nice code

File: .github/workflows/run_data_sync.yml (hunk 1 above)
Comment (Vensent) [reply .github/workflows/run_data_sync.yml:201]: > we do not support gpx right?

代码可以先放上来，现在内部消息是这个oppo heytap健康的数据暂时还没有完全对外开放，不过在计划中了。我的代码是--with-tcx、--with-gpx都可以的。需要更新一下说明不？

File: run_page/codoon_sync.py [hunk 3]
```diff
@@ -163,7 +184,12 @@ def tcx_output(fit_array, run_data):
     activity_id.text = fit_start_time  # Codoon use start_time as ID
     activity.append(activity_id)
     #   Creator
-    activity_creator = ET.Element("Creator")
+    activity_creator = ET.Element(
+        "Creator",
+        {
+            "xsi:type": "Device_t"
```
[the remark this answers is not in the document]
Comment (Vensent) [reply run_page/codoon_sync.py:190]: > 能取个更有表示点的名字么？

我打开佳明的tcx长这样：


`
<Creator xsi:type="Device_t">
  <Name>Forerunner 945</Name>
  <UnitId>********</UnitId>
  <ProductID>****</ProductID>
  <Version>
    <VersionMajor>13</VersionMajor>
    <VersionMinor>0</VersionMinor>
    <BuildMajor>0</BuildMajor>
    <BuildMinor>0</BuildMinor>
  </Version>
</Creator>
`

应该不能改成其他的样子了。

File: run_page/codoon_sync.py (hunk 2 above)
Comment (Vensent) [reply run_page/codoon_sync.py:249]: > do not need close here

python里面写文件的流最后不用关么？因为我平时用java多一点，写完流必须要关。。

Comment (Vensent) [reply run_page/codoon_sync.py:249]: got reference [here](https://stackoverflow.com/a/50712710/3998710)

File: run_page/codoon_sync.py [hunk 4]
```diff
@@ -149,7 +153,24 @@ def tcx_output(fit_array, run_data):
...
+    )
+    training_center_database.append(author)
+    author_name = ET.Element("Name")
+    author_name.text = "Connect Api"
+    author.append(author_name)
+    author_lang = ET.Element("LangID")
+    author_lang.text = "en"
+    author.append(author_lang)
+    author_part = ET.Element("PartNumber")
+    author_part.text = "006-D2449-00"
```
Comment (ben-29) [inline run_page/codoon_sync.py:172]: Extract this string to a variable and leave a comment specifying which Garmin device it is for.

**Review by ben-29 — COMMENTED** [review]
Please fix the lint using `black` to pass the CI

**Comment by Vensent** [comment]
Let me check it locally before submitting PR again.

**Comment by yihong0618** [comment]
> Let me check it locally before submitting PR again.

just continue commit is Ok~~

**Comment by Vensent** [comment]
Please review again. Thx @yihong0618 @ben-29

File: run_page/tcx_to_garmin_sync.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    tcx_files_dict = {
+        int(i[0].trackpoints[0].time.timestamp()): i[1]
+        for i in tcx_files
+        if len(i[0].trackpoints) > 0
+        and int(i[0].trackpoints[0].time.timestamp()) > last_time
+    }
+
+    dict(sorted(tcx_files_dict.items()))
+
+    return tcx_files_dict.values()
```
Comment (Vensent) [inline run_page/tcx_to_garmin_sync.py:35]: if the TCX_FOLDER is empty, we'll get an empty result. And it's OK. @ben-29

**Comment by ben-29** [comment]
LGTM

**Review by ben-29 — APPROVED** [review]

**Review by yihong0618 — APPROVED** [review]
Nice job

**Pull request merged**